runtime/src/accounts_db.rs

```diff
@@ -3396,30 +3411,90 @@ impl AccountsDb {
+    pub fn remove_unrooted_slots(&self, remove_slots: &[Slot]) {
+        let rooted_slots = self
+            .accounts_index
+            .get_rooted_from_list(remove_slots.iter());
+        if !rooted_slots.is_empty() {
+            panic!(
+                "Trying to remove accounts for rooted slots {:?}",
+                rooted_slots
+            );
+        }
+
+        let RemoveUnrootedSlotsSynchronization {
+            slots_under_contention,
+            signal,
+        } = &self.remove_unrooted_slots_synchronization;
+
+        {
+            // Slots that are currently being flushed by flush_slot_cache()
+            let mut contended_slots = slots_under_contention.lock().unwrap();
+
+            // Slots that are currently being flushed by flush_slot_cache() AND
+            // we want to remove in this function
+            let mut contended_cache_flush_slots: Vec<Slot> = remove_slots
+                .iter()
+                .filter(|remove_slot| {
+                    let is_cache_flushing_slot = contended_slots.contains(remove_slot);
+                    if !is_cache_flushing_slot {
+                        // Reserve the slots that we want to purge that aren't currently
+                        // being flushed to prevent cache from flushing those slots in
+                        // the future
+                        contended_slots.insert(**remove_slot);
+                    }
+                    // If the cache is currently flushing this slot, add it to the list
+                    is_cache_flushing_slot
+                })
+                .cloned()
+                .collect();
+
+            // Wait for cache flushes to finish
+            loop {
+                if !contended_cache_flush_slots.is_empty() {
+                    // Wait for the signal that the cache has finished flushing a slot
+                    //
+                    // Don't wait if the contended_cache_flush_slots is empty, otherwise
+                    // we may never get a signal since there's no cache flush thread to
+                    // do the signaling
+                    contended_slots = signal.wait(contended_slots).unwrap();
+                } else {
+                    // There are no slots being flushed to wait on, so it's safe to continue
+                    // to purging the slots we want to purge!
+                    break;
+                }
+
+                // For each slot the cache flush has finished, mark that we're about to start
+                // purging these slots by reserving it in `contended_slots`.
+                contended_cache_flush_slots.retain(|flushing_slot| {
+                    let is_being_flushed = contended_slots.contains(flushing_slot);
+                    if !is_being_flushed {
+                        // Mark that we're about to delete this slot now
+                        contended_slots.insert(*flushing_slot);
+                    }
+                    !is_being_flushed
```

Comment:
maybe, this condition needs to be flipped?

i mean, we need to `retain()` elements of `contended_cache_flush_slots` which have yet to be flushed (i.e. `is_being_flushed`; slots that are still `contain()`-ed  in `contended_slots` (and will be `remove()`-ed by the flusher thread later)?

Reply:
perfect. :)